Source/Diagnostics/ReducedDiags/FieldMomentum.cpp

```diff
@@ -0,0 +1,147 @@
+            ofs << "[" << c++ << "]step()";
+            ofs << m_sep;
+            ofs << "[" << c++ << "]time(s)";
+
+            // Loop over refinement levels
+            for (int lev = 0; lev < nLevel; ++lev)
+            {
+                ofs << m_sep;
+                ofs << "[" << c++ << "]";
+                ofs << "momentum_x_lev" << lev << "(kg*m/s)";
+                ofs << m_sep;
+                ofs << "[" << c++ << "]";
+                ofs << "momentum_y_lev" << lev << "(kg*m/s)";
+                ofs << m_sep;
+                ofs << "[" << c++ << "]";
+                ofs << "momentum_z_lev" << lev << "(kg*m/s)";
+            }
+
+            ofs << std::endl;
+            ofs.close();
+        }
+    }
+}
+
+void FieldMomentum::ComputeDiags (int step)
+{
+    // Check if the diags should be done
+    if (m_intervals.contains(step+1) == false)
+    {
+        return;
+    }
+
+    // Get a reference to WarpX instance
+    auto & warpx = WarpX::GetInstance();
+
+    // Get number of refinement levels
+    const auto nLevel = warpx.finestLevel() + 1;
+
+    // Loop over refinement levels
+    for (int lev = 0; lev < nLevel; ++lev)
+    {
+        // Get MultiFab data at given refinement level
+        const amrex::MultiFab & Ex = warpx.getEfield(lev, 0);
+        const amrex::MultiFab & Ey = warpx.getEfield(lev, 1);
+        const amrex::MultiFab & Ez = warpx.getEfield(lev, 2);
+        const amrex::MultiFab & Bx = warpx.getBfield(lev, 0);
+        const amrex::MultiFab & By = warpx.getBfield(lev, 1);
+        const amrex::MultiFab & Bz = warpx.getBfield(lev, 2);
+
+        // Allocate cell-centered MultiFabs
+        amrex::MultiFab Ex_cc(amrex::convert(Ex.boxArray(), amrex::IntVect::TheZeroVector()),
+                              warpx.DistributionMap(lev), 1, 0);
+
+        amrex::MultiFab Ey_cc(amrex::convert(Ey.boxArray(), amrex::IntVect::TheZeroVector()),
+                              warpx.DistributionMap(lev), 1, 0);
+
+        amrex::MultiFab Ez_cc(amrex::convert(Ez.boxArray(), amrex::IntVect::TheZeroVector()),
+                              warpx.DistributionMap(lev), 1, 0);
+
+        amrex::MultiFab Bx_cc(amrex::convert(Bx.boxArray(), amrex::IntVect::TheZeroVector()),
+                              warpx.DistributionMap(lev), 1, 0);
+
+        amrex::MultiFab By_cc(amrex::convert(By.boxArray(), amrex::IntVect::TheZeroVector()),
+                              warpx.DistributionMap(lev), 1, 0);
+
+        amrex::MultiFab Bz_cc(amrex::convert(Bz.boxArray(), amrex::IntVect::TheZeroVector()),
+                              warpx.DistributionMap(lev), 1, 0);
+
+        // Interpolate to cell-centered MultiFabs
+        CoarsenIO::Coarsen(Ex_cc, Ex, 0, 0, 1, 0);
+        CoarsenIO::Coarsen(Ey_cc, Ey, 0, 0, 1, 0);
+        CoarsenIO::Coarsen(Ez_cc, Ez, 0, 0, 1, 0);
+        CoarsenIO::Coarsen(Bx_cc, Bx, 0, 0, 1, 0);
+        CoarsenIO::Coarsen(By_cc, By, 0, 0, 1, 0);
+        CoarsenIO::Coarsen(Bz_cc, Bz, 0, 0, 1, 0);
```

Comment:
In Source/Diagnostics/ReducedDiags/FieldMaximum.cpp we interpolate to cell-centers on the fly, by calling directly `CoarsenIO::Interp` while looping over cells. However, here the sum/loop over cells is hidden inside the function `amrex::MultiFab::Dot`, implemented in [amrex/Src/Base/AMReX_MultiFab.cpp](https://github.com/AMReX-Codes/amrex/blob/a61b40464f3f16eff420d19d99cff0fa010eb2b6/Src/Base/AMReX_MultiFab.cpp#L35-L80). So I'm not sure how I could avoid the allocation of the cell-centered MultiFabs in this case.

Reply:
maybe @atmyers has an idea? :)

Reply:
I believe it's equivalent - the logic in `amrex::MultiFab::Dot` is complicated mostly because it has to handle CPU versus GPU execution; the actual dot product part is simple. In the case of `Source/Diagnostics/ReducedDiags/FieldReduction.H`, the CPU/GPU part is handled in the `reduce_op`, so I think the approach there (calling `CoarsenIO::Interp` with the proper staggering flags inside the lambda passed to the reduce_op` should work.